[8.0] ADD Putaway strategy allowing to set a location per product.

JosDeGraeve opened the pull request.

JosDeGraeve committed:
Putaway strategy allowing to set a location per product.

JosDeGraeve committed:
Added README.rst

JosDeGraeve committed:
Use proper exceptions

JosDeGraeve committed:
fix pylint errors

JosDeGraeve committed:
forgot to update __init__.py

pedrobaeza commented:
Thanks for the contribution, Jos. As `byproduct` is used in manufacturing terminology, better to name the module `stock_putaway_product` which doesn't lose the meaning. Some tests would be also interesting to put, as the introduced functionality changes standard workflows.

elicoidal commented:
what about stock_product_putaway?

elicoidal commented:
+we usually don't have hyphens in modules names.

elicoidal commented on stock_putaway_byproduct/README.rst:
```diff
@@ -0,0 +1,99 @@
+It defines a new type of putaway strategy where users can set a specific
+stock location per product.
+
+On the product form, the case,rack,loc fields are replaced with a specific
```
s/case,rack,loc/case, rack, location

elicoidal commented on stock_putaway_byproduct/README.rst:
```diff
@@ -0,0 +1,99 @@
+Installation
+============
+
+To install this module, just click the install button
```
add a period

elicoidal commented on stock_putaway_byproduct/README.rst:
```diff
@@ -0,0 +1,99 @@
+
+To configure this module, you need to:
+
+1. Go to Settings -> Configuration -> Warehouse
```
better use the "#." rst convention (but no blocker ;))

elicoidal commented on stock_putaway_byproduct/README.rst:
```diff
@@ -0,0 +1,99 @@
+:alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/153/8.0
+
+.. repo_id is available in https://github.com/OCA/maintainer-tools/blob/master/tools/repos_with_ids.txt
```
remove

elicoidal commented on stock_putaway_byproduct/README.rst:
```diff
@@ -0,0 +1,99 @@
+.. repo_id is available in https://github.com/OCA/maintainer-tools/blob/master/tools/repos_with_ids.txt
+.. branch is "8.0" for example
+
+Known issues / Roadmap
```
remove empty section

elicoidal commented on stock_putaway_byproduct/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# -*- coding: utf-8 -*-
```
- Please use simple headers in all py files following OCA templates
  https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

elicoidal commented on stock_putaway_byproduct/__openerp__.py:
```diff
@@ -0,0 +1,21 @@
+{
+    'name': 'Putaway strategy per product',
+    'summary': 'Set a product location per product',
+    'version': '8.0.0.1',
```
8.0.1.0.0

elicoidal commented on stock_putaway_byproduct/__openerp__.py:
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+{
+    'name': 'Putaway strategy per product',
+    'summary': 'Set a product location per product',
```
Set a product location and put-away strategy per product

elicoidal commented on stock_putaway_byproduct/models/product_putaway.py:
```diff
@@ -0,0 +1,75 @@
+    _sql_constraints = [(
+        'putaway_product_unique',
+        'unique(putaway_id,product_product_id)',
+        _('There can be only one fixed location per product!')
```
There can only be one fixed location per product!

elicoidal commented on stock_putaway_byproduct/wizard/stock_change_product_qty.py:
```diff
@@ -0,0 +1,33 @@
+            ])
+            if len(putaway_ids) > 1:
+                raise ValidationError(_('Error!'), _(
+                    'Multiple fixed locations for this product !!'))
```
Multiple fixed locations for this product!

elicoidal commented:
Hi @JosDeGraeve thanks a lot for this valuable contribution!

Some details but overall LGTM :+1:

pedrobaeza commented:
@elicoidal, it was a typo. I meant `stock_putaway_product`

elicoidal commented:
no problem: your suggestion is OK for me.

JosDeGraeve committed:
Renamed module, added demo data & tests

JosDeGraeve committed:
fixed some test issues, incorporated feedback

JosDeGraeve commented:
Hi,

Thank you for your feedback.  Should look good now.

pedrobaeza commented on stock_putaway_product/i18n/stock_putaway_byproduct.pot:
```diff
@@ -0,0 +1,88 @@
+# Translation of Odoo Server.
```
Remove this file, as it's automatically generated when committing before pushing to Transifex, and it can generate confusions.

pedrobaeza commented on stock_putaway_product/models/product_putaway.py:
```diff
@@ -0,0 +1,80 @@
+
+
+class FixedPutawayStrat(models.Model):
+    _name = 'stock.product.putaway.strat'
```
Why short the word `strategy` to `strat` which is not very understandable?

pedrobaeza commented on stock_putaway_product/wizard/stock_change_product_qty.py:
```diff
@@ -0,0 +1,37 @@
+                ('product_product_id', '=', product_product_id),
+                ('fixed_location_id', 'child_of', location_id),
+            ])
+            if len(putaway_ids) > 1:
```
You can make `putaway_ids.self_ensure_one()` and reduce some code.

JosDeGraeve committed:
removed pot file, use ensure_one(), use strategy instead of strat

JordiBForgeFlow commented on stock_putaway_product/README.rst:
```diff
@@ -0,0 +1,85 @@
+
+A putaway strategy can be used to ensure that incoming products will be
+stored in the location set on the product form.
+
```
I would add:

A recommended set-up is to create a separate Putaway Strategy per Warehouse. This will ensure that the same product will be placed in the appropiate location in each warehouse it is received.

carlosdauden replied:
This suggestion is attended in the next line.

JordiBForgeFlow commented on stock_putaway_product/demo/product_putaway.xml:
```diff
@@ -0,0 +1,29 @@
+        
+        <record id="product_putaway_per_product"
+                model="product.putaway">
+            <field name="name">Putaway Per Product</field>
```
This put away method will ensure that a product can only be received in one fixed location. Perhaps it would be better that in demo, this putaway strategy is called "WH - Putaway Per Product", making reference that it is intended to be used for the "WH" warehouse. 

When I tested I had to create a separate putaway strategy "Chic - Putaway Per Product" in order to be able to receive the same product in a specific location, in "Chic" warehouse as well.

cyrilgdn commented on stock_putaway_product/wizard/stock_change_product_qty.py:
```diff
@@ -0,0 +1,32 @@
+                ('product_product_id', '=', product_product_id),
+                ('fixed_location_id', 'child_of', location_id),
+            ])
+            putaway_ids.ensure_one()
```
@pedrobaeza @JosDeGraeve But ensure_one also raise ValueError if putaway_ids is empty... Not sure this is the expected behavior.

I tried, on a new database, to update quantity on a random product but a ValueError was raised.

pedrobaeza replied:
Yeah, you're right. It should be changed then if there's possible empty values.

cyrilgdn commented on stock_putaway_product/views/product.xml:
```diff
@@ -0,0 +1,43 @@
+                </xpath>
+                <xpath expr="//group[@name='store']" position="inside">
+                    <field name="product_putaway_ids"
+                           context="{'default_product_template_id': active_id }"
```
On a variant form view, active_id is  a product.product id.  So if I try to add a new storage location on a variant form, the "Product Variant" field choices contains variants of another template.

cyrilgdn commented on stock_putaway_product/models/product.py:
```diff
@@ -0,0 +1,25 @@
+        comodel_name='stock.product.putaway.strategy',
+        related='product_variant_ids.product_putaway_ids',
+        string="Product stock locations")
+
```
@JosDeGraeve After some tests (and intense reflection to understand :-)), it does not seem to work with variants...  With this related field, its impossible to create put-away strategy for two variants on same product.

Example: With demo data enable,  iPad Retina Display already has a put-away strategy for variant 16 GB, White, 2.4 GHz. If you try to create a strategy for another variant, it will raise a Constraint Error (on unique(putaway_id,product_product_id)).

The One2many widget try to add the new strategy on product.template.product_putaway_ids but with related, it seems that he write the new strategy on first variant (regardless of the product_product_id's value in putaway.strategy) .

I am a new developer Odoo, then a confirmation from an expert would be appreciated but I think you cannot write on this field...

JosDeGraeve committed:
Fix constraint error for products with multiple variants.

JosDeGraeve committed:
FIX: don't fail when no putaway_ids are found.

putaway_ids.ensure_one() fails also if no records are found.

JosDeGraeve committed:
Demo data: better reflect the link with WH warehouse

The putaway strategy is specific for a specific warehouse main location.
Better to reflect this in the naming

JosDeGraeve committed:
README: better reflect the link with WH warehouse

The putaway strategy is specific for a warehouse main location.
Better to reflect this in the naming

JosDeGraeve commented:
Hi, thanks for the extra input and @cyrilgdn thanks for finding a bug.

JosDeGraeve commented:
Is there something more needed to have this merged ?

carlosdauden reviewed (approved):
👍

JordiBForgeFlow commented:
We got three approvals. Thus merged. Thanks for this nice contribution @JosDeGraeve.

pedrobaeza commented:
Jordi, please squash on merge next time to avoid a lot of commits that doesn't add anything to the commit history.

JordiBForgeFlow commented:
OK!

On Thu, Sep 29, 2016 at 2:17 PM, Pedro M. Baeza notifications@github.com
wrote:

> Jordi, please squash on merge next time to avoid a lot of commits that
> doesn't add anything to the commit history.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> https://github.com/OCA/stock-logistics-warehouse/pull/151#issuecomment-250449932,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHU_VsqUoOGvcrtn_CAB4zNoly7S7z29ks5qu6xkgaJpZM4H5ggw
> .

## 

Jordi Ballester Alomar
CEO & Founder | Eficent
(+34) 629530707 | jordi.ballester@eficent.com | http://www.eficent.com
Twitter: https://twitter.com/jbeficent_erp | Linkedin:
https://www.linkedin.com/in/jordiballesteralomar
